plugins/modules/vault_kv2_write.py

```diff
@@ -0,0 +1,224 @@
+    HAS_HVAC = True
+
+
+def run_module():
+    argspec = HashiVaultModule.generate_argspec(
+        engine_mount_point=dict(type="str", default="secret"),
+        path=dict(type="str", required=True),
+        data=dict(type="dict", required=True),
+        cas=dict(type="int"),
+        patch=dict(type="bool", default=False),
+    )
+
+    module = HashiVaultModule(
+        argument_spec=argspec,
+        supports_check_mode=True,
+    )
+
+    if not HAS_HVAC:
+        module.fail_json(
+            msg=missing_required_lib("hvac"),
+            exception=HVAC_IMPORT_ERROR,
+        )
+
+    mount_point = module.params.get("engine_mount_point")
+    path = module.params.get("path")
+    cas = module.params.get("cas")
+    data = module.params.get("data")
+    patch = module.params.get("patch")
+
+    module.connection_options.process_connection_options()
+    client_args = module.connection_options.get_hvac_connection_options()
+    client = module.helper.get_vault_client(**client_args)
+
+    if cas and patch is True:
+        module.fail_json(msg="Cannot use cas when patch is true.")
+
+    try:
+        module.authenticator.validate()
+        module.authenticator.authenticate(client)
+    except (NotImplementedError, HashiVaultValueError) as e:
+        module.fail_json(msg=to_native(e), exception=traceback.format_exc())
+
+    try:
+        response = client.secrets.kv.v2.read_secret_version(
+            path=path, mount_point=mount_point
+        )
+        if "data" not in response or "data" not in response["data"]:
+            module.fail_json(msg="Vault response did not contain data: %s" % response)
+        current_data = response["data"]["data"]
+    except hvac.exceptions.InvalidPath:
+        if patch is True:
+            module.fail_json(
+                msg="Path '%s' has not been written to previously. Patch only works on an existing secret."
+                % path,
+                exception=traceback.format_exc(),
+            )
+        current_data = {}
+    except hvac.exceptions.Forbidden:
+        module.fail_json(
+            msg="Permission denied reading %s" % path,
+            exception=traceback.format_exc(),
+        )
+    except hvac.exceptions.VaultError:
+        module.fail_json(
+            msg="VaultError reading %s" % path,
+            exception=traceback.format_exc(),
+        )
+
+    if patch is True:
+        changed = any(current_data.get(k) != v for k, v in data.items())
+    else:
+        changed = current_data != data
+
+    if changed is True and module.check_mode is False:
+        args = {
+            "path": path,
+            "secret": data,
+            "mount_point": mount_point,
+        }
+        if cas:
+            args["cas"] = cas
+
+        try:
+            if patch is True:
+                raw = client.secrets.kv.v2.patch(**args)
+            else:
+                raw = client.secrets.kv.v2.create_or_update_secret(**args)
+        except hvac.exceptions.Forbidden:
```

Comment:
I think we might need to account for a few other exception types.
I believe we'd get `InvalidRequest` (400) if we tried to update a secret where cas is required but we don't supply it.

I'm also wondering what the exception/status code is if we supply a CAS value and it's wrong (like we gave it 2, but the latest is 1 or 3), I don't know off-hand, but we should add those scenarios to the tests.

Reply:
I'll add a test for CAS required.


Regarding a wrong cas value:
```
hvac.exceptions.InvalidRequest: check-and-set parameter did not match the current version, on post http://vault:8200/v1/kv2/data/testproject/write1
```

I have a test for it https://github.com/ansible-collections/community.hashi_vault/blob/f2d1c2fbedf819c7ec1fb39927bca6654111cb85/tests/integration/targets/module_vault_kv2_write/tasks/test.yml#L271-L283


Should I also catch `InvalidRequest`?

Reply:
I believe we catch that too, yes.